source/nsdynmemLIB/nsdynmemLIB.c

```diff
@@ -196,67 +168,43 @@ void *ns_dyn_mem_alloc(int16_t alloc_size)
         platform_enter_critical();
 
         while (moved < h_size) {
-            if (ns_sector_validate(ptr, 0) == 0) {
+            if (ns_sector_validate(ptr, direction) == 0) {
                 s_size = *ptr;
-                if (s_size == 0) {
-                    heap_failure(NS_DYN_MEM_HEAP_SECTOR_CORRUPTED);
-                    retval = 0;
-                    break;
-                } else if (s_size < 0) {
-                    //Free
-                    if (((-s_size) >= alloc_sector)) {
-                        /*found block*/
-                        //Convert Current size
-                        s_size = -s_size;
 
+                if (s_size < 0) {
+                    s_size = -s_size;
+                    if (s_size >= alloc_sector) {
+                        /*found block*/
                         if (s_size > (alloc_sector + 4)) {
-                            //debug("Bigger\r\n");
-                            //Set sector start len & Move pointer to Allocated data sector
-                            *ptr = alloc_sector;
-                            //Move Pointer  to next free place
-                            ptr -= alloc_sector;
-                            //Set return Pointer
+                            //There is enough room for new hole so create it first
+                            temp_ptr = (ptr + ((alloc_sector + 2) * direction));
+                            free_size = s_size - alloc_sector - 2;
+                            *temp_ptr = -(free_size);
+                            temp_ptr += direction * (free_size + 1);
```

Comment:
Be consistent about multiplying with direction or not. Do it everywhere or nowhere.